### PR TITLE
ci: populate caches in before script

### PR DESCRIPTION
**Pull request by haampie**
Avoid that multiple processes populate ~/.spack/cache, which sometimes
times out.

### CODE REVIEW

**Review by kwryankrattiger — DISMISSED** [review]

**Comment by haampie** [comment]
Using https://gitlab.spack.io/spack/spack/-/jobs/7604582 vs https://gitlab.spack.io/spack/spack/-/jobs/7594245

Down from 52 minutes 9 seconds to  8 minutes 53 seconds, so that's at least _something_.

So, on average 12% of a spack install process is actually spent on installing now... That's better than 1% like it used to be, but better would've been 99% :weary: 

For dependencies the sum of the time spent in `spack install` processes, including the build is `38m44s`, wall time of the CI job was `8m53s`, so yeah, there is a ~4x speedup thanks to parallel install. 

But then again... the overhead of `spack install` itself is so large that it's not enough to even justify parallel installation of dependencies :laughing:

**Comment by haampie** [comment]
My suspicion is that the rest of the time is wasted on 

1. Waiting for a database write lock
2. Reading the new database into memory once another process has written to it
3. Writing the database + 1 newly installed spec back to the filesystem.

I'm thinking it's that cause if you look at the logs, there's an increasing amount of time spent on `{built-in method time.sleep}` which is used for locks. And sometimes it drops, probably when there's a bottleneck in the DAG (e.g. a package with only one dependency)

```
ncalls  tottime  percall  cumtime  percall
 24    2.703    0.113    2.703    0.113 {built-in method time.sleep}
 23    2.503    0.109    2.503    0.109 {built-in method time.sleep}
 37    5.406    0.146    5.406    0.146 {built-in method time.sleep}
 66   12.613    0.191   12.613    0.191 {built-in method time.sleep}
 77   17.719    0.230   17.719    0.230 {built-in method time.sleep}
 89   24.526    0.276   24.526    0.276 {built-in method time.sleep}
 28    3.604    0.129    3.604    0.129 {built-in method time.sleep}
 95   27.529    0.290   27.529    0.290 {built-in method time.sleep}
107   33.535    0.313   33.535    0.313 {built-in method time.sleep}
 71   10.612    0.149   10.612    0.149 {built-in method time.sleep}
121   39.341    0.325   39.341    0.325 {built-in method time.sleep}
128   42.443    0.332   42.443    0.332 {built-in method time.sleep}
 63   11.512    0.183   11.512    0.183 {built-in method time.sleep}
139   49.551    0.356   49.551    0.356 {built-in method time.sleep}
147   53.155    0.362   53.155    0.362 {built-in method time.sleep}
  8    0.801    0.100    0.801    0.100 {built-in method time.sleep}
 25    3.003    0.120    3.003    0.120 {built-in method time.sleep}
141   41.643    0.295   41.643    0.295 {built-in method time.sleep}
132   46.048    0.349   46.048    0.349 {built-in method time.sleep}
187   73.575    0.393   73.575    0.393 {built-in method time.sleep}
149   52.955    0.355   52.955    0.355 {built-in method time.sleep}
  3    0.300    0.100    0.300    0.100 {built-in method time.sleep}
  5    0.501    0.100    0.501    0.100 {built-in method time.sleep}
 42    5.807    0.138    5.807    0.138 {built-in method time.sleep}
 54    8.610    0.159    8.610    0.159 {built-in method time.sleep}
 61    9.811    0.161    9.811    0.161 {built-in method time.sleep}
 72   15.617    0.217   15.617    0.217 {built-in method time.sleep}
 78   18.620    0.239   18.620    0.239 {built-in method time.sleep}
 80   19.621    0.245   19.621    0.245 {built-in method time.sleep}
 88   24.025    0.273   24.025    0.273 {built-in method time.sleep}
100   30.032    0.300   30.032    0.300 {built-in method time.sleep}
  6    0.601    0.100    0.601    0.100 {built-in method time.sleep}
113   34.136    0.302   34.136    0.302 {built-in method time.sleep}
114   35.837    0.314   35.837    0.314 {built-in method time.sleep}
123   40.342    0.328   40.342    0.328 {built-in method time.sleep}
137   48.551    0.354   48.551    0.354 {built-in method time.sleep}
135   45.947    0.340   45.947    0.340 {built-in method time.sleep}
 25    3.003    0.120    3.003    0.120 {built-in method time.sleep}
  3    0.300    0.100    0.300    0.100 {built-in method time.sleep}
```

Although it's surprising it takes _that_ long on an empty database... Maybe the issue is that polling is too slow, since we're starting to wait 500ms between attempts after 10 seconds of not obtaining the lock...

File: share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml [hunk 1]
```diff
@@ -210,6 +210,8 @@ spack:
     pipeline-gen:
     - build-job:
         image: ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01
+        before_script-:
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/e4s-power/spack.yaml]: Can this get moved in to the before script in the top level? We see this behavior on some other runners beyond just power.

File: var/spack/repos/builtin/packages/py-matplotlib/package.py [hunk 2]
```diff
@@ -136,6 +136,7 @@ class PyMatplotlib(PythonPackage):
     )
     variant("latex", default=False, description="Enable LaTeX text rendering support")
     variant("fonts", default=False, description="Enable support for system font detection")
+    variant("rebuildmeeeee", default=True, description="Rebuild me please")
```
Comment (kwryankrattiger) [inline var/spack/repos/builtin/packages/py-matplotlib/package.py]: Don't forget to remove this

**Comment by kwryankrattiger** [comment]
@haampie when you get a chance to clean this up I think it would be good to get it merged.

File: share/spack/gitlab/cloud_pipelines/configs/ci.yaml [hunk 3]
```diff
@@ -8,6 +8,8 @@ ci:
 
   pipeline-gen:
   - build-job:
+      before_script:
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/configs/ci.yaml]: This still needs the `-:` syntax

Comment (haampie) [inline share/spack/gitlab/cloud_pipelines/configs/ci.yaml]:
```suggestion
      before_script-:
```

File: var/spack/repos/builtin/packages/py-matplotlib/package.py [hunk 4]
```diff
@@ -137,6 +137,7 @@ class PyMatplotlib(PythonPackage):
     )
     variant("latex", default=False, description="Enable LaTeX text rendering support")
     variant("fonts", default=False, description="Enable support for system font detection")
+    variant("rebuild", default="yes", description="Test in CI")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/py-matplotlib/package.py]:
```suggestion
```

**Review by kwryankrattiger — APPROVED** [review]
LGTM, thanks for all of the benchmarking around this!

**Pull request merged**